ngraph/core/reference/include/ngraph/runtime/reference/deformable_convolution.hpp

```diff
@@ -0,0 +1,246 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/runtime/reference/convolution.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            namespace def_conv_impl
+            {
+                inline void validate_params(const Shape& in_shape,
+                                            const Shape& o_shape,
+                                            const Shape& f_shape,
+                                            const Strides& strides,
+                                            const Strides& dilations,
+                                            const CoordinateDiff& pads_begin,
+                                            const CoordinateDiff& pads_end)
+                {
+                    // this implementation supports 2D deformable convolutions
+                    NGRAPH_CHECK(in_shape.size() == 4, "Unsupported input rank: ", in_shape);
+
+                    NGRAPH_CHECK(o_shape.size() == 4, "Unsupported offset rank: ", o_shape);
+
+                    NGRAPH_CHECK(f_shape.size() == 4, "Unsupported kernel rank: ", f_shape);
+
+                    const auto spatial_dims = in_shape.size() - 2;
+                    NGRAPH_CHECK(strides.size() == spatial_dims,
+                                 "Strides not definied for all and only spatial dimensions");
+
+                    NGRAPH_CHECK(dilations.size() == spatial_dims,
+                                 "Dilations not defined for all and only spatial dimensions");
+
+                    NGRAPH_CHECK((pads_begin.size() == pads_end.size()) &&
+                                     (pads_begin.size() == spatial_dims),
+                                 "Pads not defined for all and only spatial dimensions");
+                }
+
+                inline Shape shape_reduce(const Shape& s) { return Shape(++s.begin(), s.end()); }
```

Comment:
It's not valid shape for this reference implementation so it can't be provided by user (it is checked in validate_params() at entry). I don't see compelling reason to check for it also in this function since it's only a local helper.